apps/studio/src/server/modules/resource/resource.router.ts

```diff
@@ -0,0 +1,65 @@
+import { TRPCError } from "@trpc/server"
+
+import {
+  getChildrenSchema,
+  getMetadataSchema,
+  moveSchema,
+} from "~/schemas/resource"
+import { protectedProcedure, router } from "~/server/trpc"
+import { db } from "../database"
+
+export const resourceRouter = router({
+  getMetadataById: protectedProcedure
+    .input(getMetadataSchema)
+    .query(async ({ input: { resourceId } }) => {
+      const resource = await db
+        .selectFrom("Resource")
+        .where("Resource.id", "=", String(resourceId))
+        .select(["permalink", "Resource.id", "Resource.title"])
+        .executeTakeFirst()
+
+      if (!resource) {
+        throw new TRPCError({ code: "NOT_FOUND" })
+      }
+
+      return resource
+    }),
+
+  getChildrenOf: protectedProcedure
+    .input(getChildrenSchema)
+    .query(async ({ input: { resourceId } }) => {
+      let query = db
+        .selectFrom("Resource")
+        .select(["title", "permalink", "type", "id"])
+        .orderBy("type", "asc")
+        .orderBy("title", "asc")
+
+      if (resourceId === null) {
+        query = query.where("parentId", "is", null)
+      } else {
+        query = query.where("Resource.parentId", "=", String(resourceId))
+      }
+
+      return query.execute()
+    }),
+  move: protectedProcedure
+    .input(moveSchema)
+    .mutation(async ({ input: { movedResourceId, destinationResourceId } }) => {
```

Comment:
we should validate in the backend that the destinationResourceId is a valid type (that can have resources under it).

e.g. check destination type is "Folder"

Reply:
donezxc